File: engine/effects/registry.py
"""Runtime registry for effect handlers."""

from __future__ import annotations

from typing import Callable, Dict

from engine.effects.base import Effect

EffectHandler = Callable[[Effect], dict]


class EffectRegistry:
    def __init__(self) -> None:
        self._handlers: Dict[str, EffectHandler] = {}

    def register(self, kind: str, handler: EffectHandler) -> None:
        self._handlers[kind] = handler

    def resolve(self, effect: Effect) -> dict:
        handler = self._handlers.get(effect.kind)
        if handler is None:
            return {"status": "noop", "reason": f"no handler for kind={effect.kind}"}
        return handler(effect)

File: src/effects/registry.ts
/**
 * Runtime registry for effect handlers.
 * Mirrors engine/effects/registry.py
 */

import { Effect } from "./base";

export type EffectHandler = (effect: Effect) => Record<string, unknown>;

export class EffectRegistry {
  private handlers: Map<string, EffectHandler> = new Map();

  register(kind: string, handler: EffectHandler): void {
    this.handlers.set(kind, handler);
  }

  resolve(effect: Effect): Record<string, unknown> {
    const handler = this.handlers.get(effect.kind);
    if (!handler) {
      return {
        status: "noop",
        reason: `no handler for kind=${effect.kind}`,
      };
    }
    return handler(effect);
  }
}
